File: sdk/python/src/inferlet/forward.py
"""
Forward — single forward-pass primitive with auto page management.

``ctx.forward()`` returns a :class:`Forward` builder. Attach inputs,
samplers, probes, masks, then ``await forward.execute()``::

    fwd = ctx.forward()
    fwd.input(prompt_tokens)
    h = fwd.sample([len(prompt_tokens) - 1], Sampler.argmax())
    out = await fwd.execute()
    token = out.token(h)

For prefill / scoring / custom decode loops. The :class:`Generator` layer
is built on top of this for the common token-generation case.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from wit_world.imports import inference as _inf
from wit_world.imports import media as _media
from wit_world.imports.inference import (
    SlotOutput_Distribution,
    SlotOutput_Entropy,
    SlotOutput_Logits,
    SlotOutput_Logprobs,
    SlotOutput_Token,
)

if TYPE_CHECKING:
    from .adapter import Adapter
    from .context import Context
    from .sample import Sampler


# =============================================================================
# Slot handles
# =============================================================================


@dataclass(frozen=True, slots=True)
class SampleHandle:
    """Reference to a sampler slot. Pass to :meth:`Output.token` /
    :meth:`Output.tokens_at` to read the result."""

    slot: int
    arity: int = 1


@dataclass(frozen=True, slots=True)
class ProbeHandle:
    """Reference to a probe slot. Carries the probe ``kind`` so the
    Output's accessor can sanity-check at runtime."""

    slot: int
    kind: str  # "logits" | "distribution" | "logprobs" | "entropy"


def _probe_kind(probe) -> str:
    """Map a probe spec instance to its accessor-kind string."""
    # Local import — `sample.py` is a leaf module (no upward deps), so
    # bringing it in at module load time is fine, but doing it lazily
    # here keeps `forward.py` independently importable for testing.
    from .sample import Distribution, Entropy, Logits, Logprob, Logprobs

    if isinstance(probe, Logits):
        return "logits"
    if isinstance(probe, Distribution):
        return "distribution"
    if isinstance(probe, (Logprob, Logprobs)):
        return "logprobs"
    if isinstance(probe, Entropy):
        return "entropy"
    raise TypeError(f"Unknown probe type: {type(probe).__name__}")


# =============================================================================
# Slot specs (internal)
# =============================================================================


@dataclass(slots=True)
class _SampleSlot:
    indices: list[int]
    sampler: Any  # Sampler


@dataclass(slots=True)
class _ProbeSlot:
    index: int
    probe: Any  # Logits | Distribution | Logprob | Logprobs | Entropy


# =============================================================================
# Forward
# =============================================================================


class Forward:
    """Single forward pass. Construct via :meth:`Context.forward`.

    Builder methods return ``self`` so chains compose. ``await execute()``
    runs the host call, advances the context cursor, and returns an
    :class:`Output`.
    """

    __slots__ = (
        "_ctx",
        "_auto_inputs",
        "_explicit_inputs",
        "_slots",
        "_next_slot",
        "_mask",
        "_attn_mask",
        "_adapter",
        "_zo_seed",
        "_images",
        "_audios",
        "_defer_commit",
    )

    def __init__(self, ctx: Context) -> None:
        self._ctx = ctx
        self._auto_inputs: list[int] = []
        self._explicit_inputs: list[tuple[list[int], list[int]]] = []
        self._slots: list[_SampleSlot | _ProbeSlot] = []
        self._next_slot = 0
        self._mask: list[int] | None = None
        self._attn_mask: list[list[int]] | None = None
        self._adapter: Adapter | None = None
        self._zo_seed: int | None = None
        self._images: list[tuple[_media.Image, int]] = []
        self._audios: list[tuple[_media.Audio, int]] = []
        self._defer_commit = False

    # ── Position accessors ────────────────────────────────────────────

    def start_position(self) -> int:
        """Position the *first* auto-input token will occupy. Equal to the
        owning context's ``seq_len`` at the time ``forward()`` was called.

        The sampler at index ``i`` (when ``forward.sample([i], ...)``)
        lands at ``start_position() + i``.
        """
        return self._ctx._seq_len

    def page_size(self) -> int:
        """Tokens per KV page for the owning context."""
        return self._ctx._page_size

    def defer_commit(self) -> Forward:
        """Run the pass without advancing the context cursor afterward."""
        self._defer_commit = True
        return self

    # ── Inputs ────────────────────────────────────────────────────────

    def input(self, tokens: list[int]) -> Forward:
        """Append ``tokens`` at positions starting at the context's current
        sequence length. Multiple calls accumulate. After ``execute()``
        these tokens occupy KV slots and ``seq_len`` advances."""
        self._auto_inputs.extend(tokens)
        return self

    def input_at(self, tokens: list[int], positions: list[int]) -> Forward:
        """Feed ``tokens`` at caller-supplied ``positions``. Use for
        scoring at arbitrary positions (e.g. multi-candidate evaluation).
        These tokens are NOT auto-committed — caller manages page
        bookkeeping if positions overlap or extend beyond ``seq_len``."""
        if len(tokens) != len(positions):
            raise ValueError("tokens and positions must be the same length")
        self._explicit_inputs.append((list(tokens), list(positions)))
        return self

    # ── Slot attach ───────────────────────────────────────────────────

    def sample(self, indices: list[int], sampler: Sampler) -> SampleHandle:
        """Attach a sampler at one or more ``indices`` (0-based into the
        auto-input window). Returns a handle for reading the sampled
        token(s) on the resulting :class:`Output`.

        A multi-arity sampler produces ``len(indices)`` Token slots in
        the output, so the next slot index advances by that count — any
        subsequent ``sample`` / ``probe`` call sees the right offset.
        """
        arity = len(indices)
        h = SampleHandle(slot=self._next_slot, arity=arity)
        self._slots.append(_SampleSlot(list(indices), sampler))
        self._next_slot += arity
        return h

    def probe(self, index: int, probe: Any) -> ProbeHandle:
        """Attach a probe at a single ``index``. Returns a handle whose
        ``kind`` selects which ``output.*`` accessor decodes the result."""
        h = ProbeHandle(slot=self._next_slot, kind=_probe_kind(probe))
        self._slots.append(_ProbeSlot(index, probe))
        self._next_slot += 1
        return h

    # ── Decoration ────────────────────────────────────────────────────

    def mask(self, brle: list[int]) -> Forward:
        """Set a static logit mask (BRLE) applied at every sampled position."""
        self._mask = list(brle)
        return self

    def attention_mask(self, masks: list[list[int]]) -> Forward:
        """Set per-query-position attention masks. Length must match the
        total number of query positions across all ``input`` /
        ``input_at`` calls. If unset, the runtime synthesizes a causal mask."""
        self._attn_mask = [list(m) for m in masks]
        return self

    def adapter(self, adapter: Adapter) -> Forward:
        """Apply an adapter (LoRA, etc.) for this forward pass."""
        self._adapter = adapter
        return self

    def zo_seed(self, seed: int) -> Forward:
        """Set a zo (Evolution Strategies) seed for this forward pass."""
        self._zo_seed = seed
        return self

    def input_image(self, image: _media.Image, anchor: int) -> Forward:
        """Splice an encoded visual span at sequence position ``anchor``."""
        self._images.append((image, anchor))
        return self

    def input_audio(self, audio: _media.Audio, anchor: int) -> Forward:
        """Splice an encoded audio span at sequence position ``anchor``."""
        self._audios.append((audio, anchor))
        return self

    # ── Execute ───────────────────────────────────────────────────────

    async def execute(self) -> Output:
        """Run the forward pass with explicit KV descriptors.

        Raises ``ValueError`` if no inputs and no slots are attached —
        a vacuous Forward almost always indicates a missed
        ``input(...)`` or ``sample(...)`` call.
        """
        ctx = self._ctx
        n_auto = len(self._auto_inputs)
        n_total = n_auto + sum(len(t) for t, _ in self._explicit_inputs)

        if n_total == 0 and not self._slots and not self._images and not self._audios:
            raise ValueError(
                "Forward.execute() called with no inputs and no slots. "
                "Attach at least one input (`forward.input(...)`) or "
                "slot (`forward.sample(...)` / `forward.probe(...)`) "
                "before executing."
            )

        soft_tokens = sum(image.token_count() for image, _ in self._images)
        soft_tokens += sum(audio.token_count() for audio, _ in self._audios)
        n_write = n_auto + soft_tokens

        # Build forward pass.
        fwd = _inf.ForwardPass()
        if n_write > 0:
            generation, indices, valid_lens, ctx_pages = ctx._prepare_write(n_write)
            ctx._attach_kv(fwd, generation, indices, valid_lens, ctx_pages)
        else:
            ctx._attach_full_context(fwd)

        for image, anchor in self._images:
            fwd.input_image(image, anchor)
        for audio, anchor in self._audios:
            fwd.input_audio(audio, anchor)

        if self._adapter is not None:
            fwd.adapter(self._adapter._handle)
        if self._zo_seed is not None:
            from wit_world.imports import zo as _zo

            _zo.adapter_seed(fwd, self._zo_seed)

        if n_auto > 0:
            positions = list(range(ctx._seq_len, ctx._seq_len + n_auto))
            fwd.input_tokens(self._auto_inputs, positions)
        for tokens, positions in self._explicit_inputs:
            fwd.input_tokens(tokens, positions)

        # Slot attaches go in declaration order — slot indices match
        # what we handed back via SampleHandle / ProbeHandle.
        for spec in self._slots:
            if isinstance(spec, _SampleSlot):
                fwd.sampler(spec.indices, spec.sampler._variant)
            else:  # _ProbeSlot
                fwd.sampler([spec.index], spec.probe._to_wit())

        if self._mask is not None:
            fwd.logit_mask(self._mask)
        if self._attn_mask is not None:
            fwd.attention_mask(self._attn_mask)

        raw = await fwd.execute()

        if n_write > 0 and not self._defer_commit:
            ctx._seq_len += n_write
            ctx._history.extend(self._auto_inputs)
        if soft_tokens > 0:
            ctx._snapshottable = False

        return Output(raw)


# =============================================================================
# Output
# =============================================================================


class Output:
    """Result of one forward-pass execution — produced by both
    :meth:`Forward.execute` and :meth:`GenStep.execute`.

    **Common path** (Generator): read :attr:`tokens` for the accepted
    tokens this step (post stop / max-tokens truncation).

    **Raw Forward**: read sampler slots via :meth:`token` / :meth:`tokens_at`
    using handles returned at attach time. The :attr:`tokens` field is empty.

    **Probes** (both paths): :meth:`distribution` / :meth:`logits` /
    :meth:`logprobs` / :meth:`entropy` take a :class:`ProbeHandle`.

    Mismatched access (reading a sampler slot through a probe handle, or
    vice versa) returns ``None``.
    """

    __slots__ = ("_raw", "tokens", "auto_sampler")

    def __init__(
        self,
        raw: _inf.Output,
        tokens: list[int] | None = None,
        auto_sampler: SampleHandle | None = None,
    ) -> None:
        self._raw = raw
        #: Generator-accepted tokens this step, post stop / max-tokens
        #: truncation. Empty for raw ``Forward.execute()`` (no Generator state).
        self.tokens: list[int] = tokens if tokens is not None else []
        #: Handle for the Generator's auto-attached sampler. ``None`` for
        #: raw Forward results and for steps where ``clear_sampler()``
        #: was called on the GenStep.
        self.auto_sampler: SampleHandle | None = auto_sampler

    @property
    def raw(self) -> _inf.Output:
        """Underlying WIT output (slot list + speculative side channel)."""
        return self._raw

    # ── Sampler accessors ────────────────────────────────────────────

    def token(self, h: SampleHandle) -> int | None:
        """First token from a single-index sampler slot."""
        slot = self._slot(h.slot)
        if isinstance(slot, SlotOutput_Token):
            return slot.value
        return None

    def tokens_at(self, h: SampleHandle) -> list[int]:
        """Tokens at the slot range a multi-index sampler covers. In
        speculative mode the list may be shorter than ``arity`` if the
        verifier rejected drafts."""
        out: list[int] = []
        for i in range(h.arity):
            slot = self._slot(h.slot + i)
            if isinstance(slot, SlotOutput_Token):
                out.append(slot.value)
            else:
                break
        return out

    # ── Probe accessors ──────────────────────────────────────────────

    def distribution(self, h: ProbeHandle) -> tuple[list[int], list[float]] | None:
        """Distribution as ``(ids, probs)`` for a :class:`Distribution` probe."""
        slot = self._slot(h.slot)
        if isinstance(slot, SlotOutput_Distribution):
            ids, probs = slot.value
            return list(ids), list(probs)
        return None

    def logits(self, h: ProbeHandle) -> bytes | None:
        """Raw logits bytes for a :class:`Logits` probe (length
        ``vocab_size * 4``, native-endian f32)."""
        slot = self._slot(h.slot)
        if isinstance(slot, SlotOutput_Logits):
            return bytes(slot.value)
        return None

    def logprobs(self, h: ProbeHandle) -> list[float] | None:
        """Logprob list for a :class:`Logprob` / :class:`Logprobs` probe.
        Length is 1 for a single-token query, K for a list query."""
        slot = self._slot(h.slot)
        if isinstance(slot, SlotOutput_Logprobs):
            return list(slot.value)
        return None

    def entropy(self, h: ProbeHandle) -> float | None:
        """Entropy for an :class:`Entropy` probe."""
        slot = self._slot(h.slot)
        if isinstance(slot, SlotOutput_Entropy):
            return slot.value
        return None

    def _slot(self, idx: int):
        if 0 <= idx < len(self._raw.slots):
            return self._raw.slots[idx]
        return None

File: sdk/rust/inferlet/src/forward.rs
//! `Forward` — a single forward-pass primitive with automatic page management.
//!
//! `Forward` wraps the WIT `forward-pass` resource and folds in the page math
//! that every consumer would otherwise repeat: working-page reservation
//! before submission, position derivation, page commit after execution.
//! Slots are attached via [`Forward::sample`] / [`Forward::probe`] and read back
//! through typed handles on the returned [`Output`].
//!
//! For the rare case where the caller needs full WIT control (custom
//! attention masks, manual page management for speculation rollback, etc.),
//! the underlying `forward-pass` resource is still available via
//! `inferlet::inference::ForwardPass`.

use std::marker::PhantomData;

use crate::ForwardPassExt;
use crate::Result;
use crate::adapter::Adapter;
use crate::context::Context;
use crate::pie::core::inference::Output as RawOutput;
use crate::pie::core::inference::{ForwardPass, SlotOutput};
use crate::sample::{self, Probe, Sampler};

// =============================================================================
// Slot handles
// =============================================================================

/// Opaque handle to a sampler slot, returned by [`Forward::sample`]. Pass to
/// [`Output::token`] / [`Output::tokens`] to read the result.
#[derive(Copy, Clone, Debug)]
pub struct SampleHandle {
    slot: u32,
    /// Number of positions the sampler was attached to. Lets `Output::tokens`
    /// read back the right window without separate bookkeeping.
    arity: u32,
}

impl SampleHandle {
    pub(crate) fn new(slot: u32, arity: u32) -> Self {
        Self { slot, arity }
    }
    pub(crate) fn slot(&self) -> u32 {
        self.slot
    }
    pub(crate) fn arity(&self) -> u32 {
        self.arity
    }
}

/// Phantom-typed handle to a probe slot, returned by [`Forward::probe`]. The
/// type parameter selects which `Output::*` accessor compiles.
#[derive(Debug)]
pub struct ProbeHandle<P> {
    slot: u32,
    _kind: PhantomData<fn() -> P>,
}

// `PhantomData<fn() -> P>` is `Copy + Clone` regardless of `P`, but the
// auto-derive generates `where P: Copy/Clone` bounds which surprise
// callers using probe markers like `Logprobs(Vec<u32>)`. Hand-roll the
// impls so handles are always `Copy`.
impl<P> Copy for ProbeHandle<P> {}
impl<P> Clone for ProbeHandle<P> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<P> ProbeHandle<P> {
    pub(crate) fn new(slot: u32) -> Self {
        Self {
            slot,
            _kind: PhantomData,
        }
    }
    pub(crate) fn slot(&self) -> u32 {
        self.slot
    }
}

// =============================================================================
// Pass
// =============================================================================

/// One slot attach, kept until `execute` so the WIT calls land in the same
/// order the handles were handed out. Probes go through `into_wit()` at
/// construction time and stash the resulting variant here.
enum SlotSpec {
    Sample {
        indices: Vec<u32>,
        sampler: Sampler,
    },
    Probe {
        index: u32,
        wit: crate::pie::core::inference::Sampler,
    },
}

/// Builder for a single forward pass. Construct via
/// [`Context::pass`](crate::Context::pass).
///
/// Builder methods return `&mut Self` so chains compose. `execute(self)` runs
/// the host call, commits pages, and returns an [`Output`].
pub struct Forward<'ctx> {
    ctx: &'ctx mut Context,
    /// Tokens fed at auto-derived positions (appended at the next free slot).
    auto_inputs: Vec<u32>,
    /// Tokens fed at caller-supplied positions. Used for non-contiguous
    /// scoring (e.g. validation) where positions don't extend `seq_len`.
    explicit_inputs: Vec<(Vec<u32>, Vec<u32>)>,
    /// Slot attachments in declaration order.
    slots: Vec<SlotSpec>,
    next_slot: u32,
    mask: Option<Vec<u32>>,
    attn_mask: Option<Vec<Vec<u32>>>,
    adapter: Option<&'ctx Adapter>,
    zo_seed: Option<i64>,
    /// Visual spans spliced at the given anchor positions, in declaration
    /// order. Emitted as `input-image` calls at execute time.
    images: Vec<(&'ctx crate::media::Image, u32)>,
    /// Audio clips spliced at the given anchor positions, in declaration
    /// order. Emitted as `input-audio` calls at execute time.
    audios: Vec<(&'ctx crate::media::Audio, u32)>,
    /// When true, `execute` reserves pages and runs the pass but does NOT
    /// commit the newly-filled working pages. Used by manual speculative
    /// loops that must keep all written tokens in *working* pages so a
    /// follow-up [`Context::truncate`] can roll back the rejected suffix —
    /// committing first (the default) would lock unverified drafts into
    /// committed KV that `truncate` cannot reach. The caller commits the
    /// verified prefix itself once acceptance is known.
    defer_commit: bool,
}

impl<'ctx> Forward<'ctx> {
    /// Position the *first* auto-input token will occupy. Equals the
    /// owning context's `seq_len` at the time `pass()` was called. Useful
    /// for callers that need to derive an attention mask or condition on
    /// the upcoming position before `execute`. The sampler at index `i`
    /// (when `pass.sample(&[i], …)`) lands at `start_position() + i`.
    pub fn start_position(&self) -> u32 {
        self.ctx.seq_len()
    }

    /// Page size of the owning context — for callers that need to size
    /// per-position structures (masks, etc.) without re-querying.
    pub fn page_size(&self) -> u32 {
        self.ctx.page_size()
    }

    /// Construct a `Forward` against a context. Prefer [`Context::pass`].
    pub(crate) fn new(ctx: &'ctx mut Context) -> Self {
        Self {
            ctx,
            auto_inputs: Vec::new(),
            explicit_inputs: Vec::new(),
            slots: Vec::new(),
            next_slot: 0,
            mask: None,
            attn_mask: None,
            adapter: None,
            zo_seed: None,
            images: Vec::new(),
            audios: Vec::new(),
            defer_commit: false,
        }
    }

    /// Skip the post-execute page commit, keeping every written token in
    /// *working* pages. The pass still reserves pages and writes KV, and the
    /// context's `seq_len` / `working_tokens` advance, but no working page is
    /// committed.
    ///
    /// Required for correct manual speculative decoding (draft + verify +
    /// truncate). The default auto-commit promotes full working pages into
    /// committed KV as soon as they fill — but a speculative verify pass
    /// writes anchor + draft tokens whose acceptance is not yet known. A
    /// committed page can no longer be rolled back by [`Context::truncate`],
    /// and committing pages mid-speculation also perturbs the KV the
    /// remaining decode attends to. Keeping the verified prefix in working
    /// pages (and letting `truncate` drop the rejected suffix) is what makes
    /// the generated sequence independent of the draft length, as lossless
    /// speculative decoding requires.
    pub fn defer_commit(&mut self) -> &mut Self {
        self.defer_commit = true;
        self
    }

    // ── Inputs ─────────────────────────────────────────────────────────

    /// Append `tokens` at positions starting at the context's current
    /// sequence length. Multiple calls accumulate. After `execute()`, these
    /// tokens occupy KV slots and the context's `seq_len` advances.
    pub fn input(&mut self, tokens: &[u32]) -> &mut Self {
        self.auto_inputs.extend_from_slice(tokens);
        self
    }

    /// Feed `tokens` at caller-supplied `positions`. Use for scoring at
    /// arbitrary positions (e.g. log-probability evaluation across a
    /// candidate window). These tokens are NOT auto-committed — the caller
    /// is responsible for any page bookkeeping if positions overlap or
    /// extend beyond `seq_len`.
    ///
    /// Panics if `tokens.len() != positions.len()`.
    pub fn input_at(&mut self, tokens: &[u32], positions: &[u32]) -> &mut Self {
        assert_eq!(
            tokens.len(),
            positions.len(),
            "input_at: tokens and positions must be the same length"
        );
        self.explicit_inputs
            .push((tokens.to_vec(), positions.to_vec()));
        self
    }

    // ── Slot attach ────────────────────────────────────────────────────

    /// Attach a sampler at one or more `indices`. Indices are 0-based into
    /// the auto-input window: `0` is the first auto-input token, `n-1` the
    /// last. Returns a handle for reading the sampled token(s) on the
    /// resulting [`Output`].
    pub fn sample(&mut self, indices: &[u32], sampler: Sampler) -> SampleHandle {
        let arity = indices.len() as u32;
        let h = SampleHandle::new(self.next_slot, arity);
        self.slots.push(SlotSpec::Sample {
            indices: indices.to_vec(),
            sampler,
        });
        // Multi-arity samplers produce `arity` Token slots in the output
        // (one per index); advance the slot index by that count so any
        // subsequent `sample` / `probe` call sees the right offset.
        self.next_slot += arity;
        h
    }

    /// Attach a probe at a single `index`. Returns a typed handle whose
    /// type parameter selects which `Output::*` accessor compiles.
    pub fn probe<P: Probe>(&mut self, index: u32, probe: P) -> ProbeHandle<P::Out> {
        let h = ProbeHandle::new(self.next_slot);
        self.slots.push(SlotSpec::Probe {
            index,
            wit: probe.into_wit(),
        });
        self.next_slot += 1;
        h
    }

    // ── Decoration ─────────────────────────────────────────────────────

    /// Set a static logit mask (BRLE) applied at every sampled position.
    pub fn mask(&mut self, brle: &[u32]) -> &mut Self {
        self.mask = Some(brle.to_vec());
        self
    }

    /// Set per-query-position attention masks. Length must match the total
    /// number of query positions across all `input` / `input_at` calls.
    /// If unset, the runtime synthesizes a causal mask.
    pub fn attention_mask(&mut self, masks: &[Vec<u32>]) -> &mut Self {
        self.attn_mask = Some(masks.to_vec());
        self
    }

    /// Apply an adapter (LoRA, etc.) for this forward pass.
    pub fn adapter(&mut self, adapter: &'ctx Adapter) -> &mut Self {
        self.adapter = Some(adapter);
        self
    }

    /// Splice an encoded visual span (image or video clip) at sequence
    /// position `anchor`. The driver runs the vision encoder and scatters the
    /// projected rows into the hidden state. Advance your position cursor by
    /// `image.position_span()` for any tokens that follow. See MULTIMODAL.md.
    pub fn input_image(&mut self, image: &'ctx crate::media::Image, anchor: u32) -> &mut Self {
        self.images.push((image, anchor));
        self
    }

    /// Splice an encoded audio clip at sequence position `anchor`. The driver
    /// runs the gemma4_audio encoder and scatters the projected rows into the
    /// hidden state. Advance your position cursor by `audio.position_span()`
    /// for any tokens that follow. See audio_frontend.md.
    pub fn input_audio(&mut self, audio: &'ctx crate::media::Audio, anchor: u32) -> &mut Self {
        self.audios.push((audio, anchor));
        self
    }

    /// Set a `zo` (Evolution Strategies) seed for this forward pass.
    pub fn zo_seed(&mut self, seed: i64) -> &mut Self {
        self.zo_seed = Some(seed);
        self
    }

    // ── Execute ────────────────────────────────────────────────────────

    /// Run the forward pass. Reserves working pages for any auto-inputs,
    /// submits all attached inputs and slots, awaits the host, commits any
    /// newly-filled pages, and updates the context's cached state.
    pub async fn execute(self) -> Result<Output> {
        let Forward {
            ctx,
            auto_inputs,
            explicit_inputs,
            slots,
            next_slot: _,
            mask,
            attn_mask,
            adapter,
            zo_seed,
            images,
            audios,
            defer_commit,
        } = self;

        let n_auto = auto_inputs.len() as u32;

        // Reserve working pages for auto-inputs (those occupy KV slots and
        // commit on the way out). Explicit inputs are scoring-only — the
        // caller manages their pages.
        if n_auto > 0 {
            let total_after = ctx.working_tokens + n_auto;
            let pages_needed = (total_after + ctx.page_size - 1) / ctx.page_size;
            let additional = pages_needed.saturating_sub(ctx.working_pages);
            if additional > 0 {
                ctx.inner
                    .reserve_working_pages(additional)
                    .map_err(|e| format!("Forward::execute: reserve_working_pages: {e}"))?;
                ctx.working_pages = pages_needed;
            }
        }

        let pass = ForwardPass::new(&ctx.model);
        pass.context(&ctx.inner);

        for (image, anchor) in images {
            pass.input_image(image, anchor);
        }
        for (audio, anchor) in audios {
            pass.input_audio(audio, anchor);
        }

        if let Some(a) = adapter {
            pass.adapter(a);
        }
        if let Some(seed) = zo_seed {
            crate::pie::zo::zo::adapter_seed(&pass, seed);
        }

        if n_auto > 0 {
            let positions: Vec<u32> = (ctx.seq_len..ctx.seq_len + n_auto).collect();
            pass.input_tokens(&auto_inputs, &positions);
        }
        for (toks, pos) in &explicit_inputs {
            pass.input_tokens(toks, pos);
        }

        for spec in slots {
            match spec {
                SlotSpec::Sample { indices, sampler } => {
                    let wit: crate::pie::core::inference::Sampler = sampler.into();
                    pass.sampler(&indices, &wit);
                }
                SlotSpec::Probe { index, wit } => {
                    pass.sampler(&[index], &wit);
                }
            }
        }

        if let Some(m) = mask {
            pass.logit_mask(&m);
        }
        if let Some(m) = attn_mask {
            pass.attention_mask(&m);
        }

        let raw = pass
            .execute_async()
            .await
            .map_err(|e| format!("Forward::execute: forward pass failed: {e}"))?;

        // Account for the newly-written KV. By default we also commit any
        // pages the auto-input tokens fully filled. When `defer_commit` is
        // set (manual speculation), we keep every written token in *working*
        // pages so a later `truncate` can roll back rejected drafts — the
        // caller commits the verified prefix afterwards.
        if n_auto > 0 {
            let new_working = ctx.working_tokens + n_auto;
            let to_commit = if defer_commit {
                0
            } else {
                new_working / ctx.page_size
            };
            if to_commit > 0 {
                ctx.inner
                    .commit_working_pages(to_commit)
                    .map_err(|e| format!("Forward::execute: commit_working_pages: {e}"))?;
            }
            ctx.committed_pages += to_commit;
            ctx.working_pages -= to_commit;
            ctx.working_tokens = new_working - to_commit * ctx.page_size;
            ctx.seq_len += n_auto;
        }

        Ok(Output::new(raw))
    }
}

// =============================================================================
// Output
// =============================================================================

/// Result of one forward-pass execution — produced by both
/// [`Forward::execute`] and [`GenStep::execute`](crate::generation::GenStep::execute).
///
/// **Common path** ([`Generator`](crate::generation::Generator)): read
/// the [`tokens`](Self::tokens) field for the accepted tokens this step
/// (post stop / max-tokens truncation). The auto-attached sampler's
/// [`SampleHandle`] is also exposed via [`auto_sampler`](Self::auto_sampler)
/// for callers that want the pre-truncation verifier output.
///
/// **Raw `Forward`**: read sampler slots via [`token`](Self::token) /
/// [`tokens_at`](Self::tokens_at) using the handles returned at attach
/// time. The [`tokens`](Self::tokens) field is empty.
///
/// **Probes** (both paths): [`distribution`](Self::distribution) /
/// [`logits`](Self::logits) / [`logprobs`](Self::logprobs) /
/// [`entropy`](Self::entropy) take a typed [`ProbeHandle`].
///
/// Mismatched access (reading a sampler slot through a probe handle, or
/// vice versa) returns `None`.
pub struct Output {
    raw: RawOutput,
    /// Generator-accepted tokens this step, post stop / max-tokens
    /// truncation. Empty for raw `Forward::execute` (no Generator state).
    pub tokens: Vec<u32>,
    auto_sampler: Option<SampleHandle>,
}

impl Output {
    pub(crate) fn new(raw: RawOutput) -> Self {
        Self {
            raw,
            tokens: Vec::new(),
            auto_sampler: None,
        }
    }

    pub(crate) fn from_generator(
        raw: RawOutput,
        tokens: Vec<u32>,
        auto_sampler: Option<SampleHandle>,
    ) -> Self {
        Self {
            raw,
            tokens,
            auto_sampler,
        }
    }

    /// Underlying WIT output, for callers who need the raw slot list or the
    /// speculative side channel (`raw.spec_tokens`, `raw.spec_positions`).
    pub fn raw(&self) -> &RawOutput {
        &self.raw
    }

    /// Handle for the generator's auto-attached sampler. `None` for raw
    /// `Forward` results and for steps where
    /// [`GenStep::clear_sampler`](crate::generation::GenStep::clear_sampler)
    /// was called.
    pub fn auto_sampler(&self) -> Option<SampleHandle> {
        self.auto_sampler
    }

    /// First token from a single-index sampler slot.
    pub fn token(&self, h: SampleHandle) -> Option<u32> {
        match self.raw.slots.get(h.slot() as usize)? {
            SlotOutput::Token(t) => Some(*t),
            _ => None,
        }
    }

    /// Tokens at the slot range a multi-index sampler covers. Returns one
    /// token per index the sampler was attached to. In speculative mode
    /// the slice may be shorter than `arity` if the verifier rejected
    /// drafts.
    pub fn tokens_at(&self, h: SampleHandle) -> Vec<u32> {
        let mut out = Vec::with_capacity(h.arity() as usize);
        let start = h.slot() as usize;
        for i in 0..(h.arity() as usize) {
            match self.raw.slots.get(start + i) {
                Some(SlotOutput::Token(t)) => out.push(*t),
                _ => break,
            }
        }
        out
    }

    /// Distribution as `(ids, probs)` for a [`Distribution`](sample::Distribution) probe.
    pub fn distribution(&self, h: ProbeHandle<sample::Distribution>) -> Option<(&[u32], &[f32])> {
        match self.raw.slots.get(h.slot() as usize)? {
            SlotOutput::Distribution((ids, ps)) => Some((ids.as_slice(), ps.as_slice())),
            _ => None,
        }
    }

    /// Raw logits bytes for a [`Logits`](sample::Logits) probe.
    pub fn logits(&self, h: ProbeHandle<sample::Logits>) -> Option<&[u8]> {
        match self.raw.slots.get(h.slot() as usize)? {
            SlotOutput::Logits(b) => Some(b.as_slice()),
            _ => None,
        }
    }

    /// Logprob list for a [`Logprob`](sample::Logprob) or
    /// [`Logprobs`](sample::Logprobs) probe. Length is 1 for a single-token
    /// query, K for a list query.
    pub fn logprobs(&self, h: ProbeHandle<sample::Logprobs>) -> Option<&[f32]> {
        match self.raw.slots.get(h.slot() as usize)? {
            SlotOutput::Logprobs(v) => Some(v.as_slice()),
            _ => None,
        }
    }

    /// Entropy for an [`Entropy`](sample::Entropy) probe.
    pub fn entropy(&self, h: ProbeHandle<sample::Entropy>) -> Option<f32> {
        match self.raw.slots.get(h.slot() as usize)? {
            SlotOutput::Entropy(v) => Some(*v),
            _ => None,
        }
    }
}
